Leaflet.js total refactor!

GoToLoop opened the pull request:
After taking a hard look at Leaflet.js (also TileMap.js) in order to find a more proper solution for some1 from the Processing's forum:
https://Forum.Processing.org/two/discussion/26239/p5-js-mappa-js-unable-to-place-map-in-parent-div#Item_4

I've found out this class _**Leaflet**_ got some tiny minor cosmetic issues. And that I could refactor it.  :P  

1) Current latest Leaflet.JS' version is now 1.3.1. But this class is loading 1.3.0 instead.
So I've changed 'https://unpkg.com/leaflet@1.3.0/dist/leaflet.js' to 'https://unpkg.com/leaflet@1.3/dist/leaflet.js'.
`@1.3`, rather than `@1.3.0`, will automatically pick latest version from range v1.3.0 to v1.3.9.
I don't think patch updates are that dangerous, are they?

2) In **canvasOverlay()**, you've got an custom subclass from _**L.Layer**_ named as _**L.overlay**_.
However, according to JS/Java's conventions, class names should be capitalized.
Therefore I've renamed _**L.overlay**_ to _**L.Layer.Overlay**_.
Then made a custom **L.overlay()** function which instantiates that just renamed subclass.  ^_^
Given the now wrapper function got the same name as the previous subclass' name, no API call had been changed!

3) Besides that renaming and a new wrapper function in its place, I've found some instructions about extending class _**L.Layer**_ below:
http://LeafletJS.com/examples/extending/extending-2-layers.html
However, your approach differs greatly from the article above.
You use an arrow function for _**L.Layer**_::**onAdd()** method, permanently sealing its `this` to that of subclass _**Leaflet**_'s.
You do that in order to access _this.canvas_ inside _**L.Layer**_::**onAdd()**, I know.
Well, I've refactored it to access _this.canvas_ as a closure variable simply called _canvas_ there.

Well, those are my main changes. There are many more tiny 1s throughout the file though.  :D  
Got no idea how much you're gonna like them. Just tell me which 1s to revert back then.

P.S.: Not tested at all!!!

GoToLoop committed:
Leaflet.js total refactor!

After taking a hard look at Leaflet.js (also TileMap.js) in order to find a more proper solution for some1 from the Processing's forum:
https://Forum.Processing.org/two/discussion/26239/p5-js-mappa-js-unable-to-place-map-in-parent-div#Item_4

I've found out this class _**Leaflet**_ got some tiny minor cosmetic issues. And I could refactor it.  :P  

1) Current latest Leaflet.JS' version is now 1.3.1. But this class is loading 1.3.0 instead.
So I've changed 'https://unpkg.com/leaflet@1.3.0/dist/leaflet.js' to 'https://unpkg.com/leaflet@1.3/dist/leaflet.js'.
`@1.3`, rather than `@1.3.0`, will automatically pick latest version from range v1.3.0 to v1.3.9.
I don't think patch updates are that dangerous, are they?

2) In **canvasOverlay()**, you've got an custom subclass from _**L.Layer**_ named as _**L.overlay**_.
However, according to JS/Java's conventions, class names should be capitalized.
Therefore I've renamed _**L.overlay**_ to _**L.Layer.Overlay**_.
Then made a custom **L.overlay()** function which instantiates that just renamed subclass.  ^_^
Given the now wrapper function got the same name as the previous subclass' name, no API call had been changed!

3) Besides that renaming and a new wrapper function in its place, I've found some instructions about extending class _**L.Layer**_ below:
http://LeafletJS.com/examples/extending/extending-2-layers.html

However, your approach differs greatly from the article above.
You use an arrow function for _**L.Layer**_::**onAdd()** method, permanently sealing its `this` to that of subclass _**Leaflet**_'s.
You do that in order to access _this.canvas_ inside _**L.Layer**_::**onAdd()**.
Well, I've refactored it to access _this.canvas_ as a closure variable simply called _canvas_ there.

Those are my main changes. There are many more tiny 1s throughout the file though.  :D  
Got no idea how much you're gonna like them. Just tell me which 1s to revert back then.

GoToLoop committed:
Update Leaflet.js

GoToLoop committed:
Update Leaflet.js

cvalenzuela commented on src/providers/tile/Leaflet.js:
```diff
@@ -1,120 +1,109 @@
-      this.loadSrc();
-    }
+
+    this.constructor.name === 'Leaflet' && this.loadSrc();
```
I am using regular `if` statements as a default instead of ternary or short-circuit evaluations. Just as a style guide.

cvalenzuela commented on src/providers/tile/Leaflet.js:
```diff
@@ -1,120 +1,109 @@
-        overlayPane.appendChild(container);
+    L.Layer.Overlay = L.Layer.extend({
+      onAdd(map) {
+        this._container = L.DomUtil.create('div', 'leaflet-layer');
```
Is there any need to have `_container` as class property? can it be just `const container`?

GoToLoop replied:
As I've stated, I'm following the tutorial from here:
http://LeafletJS.com/examples/extending/extending-2-layers.html

Property **_container** is needed for implementing _**Layer**_::**onRemove()**, so it knows which _**HTMLElement**_, which is created in _**Layer**_::**onAdd()**, to remove.

And according to Leaflet.js' API, _**Layer**_::**onRemove()** isn't optional, just like _**Layer**_::**onAdd()**:
http://LeafletJS.com/reference-1.3.0.html#layer-onremove

GoToLoop replied:
Besides, **_container** gets `delete` later on inside _**Layer**_::**onRemove()** anywayz.  ^_^

cvalenzuela replied:
perfect!

GoToLoop replied:
Just 1 extra observation: property **_container** is being temporarily added to the _**L.Layer**_ subclass, not _**Leaflet**_.

Recall that _**Layer**_::**onAdd()** isn't a fat arrow function, and its `this` isn't bound to _**Leaflet**_ anymore.

B/c the subclass of _**L.Layer**_ isn't exposed, getting access to _**L.Layer**_::*_container* isn't easy.

Although not impossible. It's possible to call _**L.Map**_::**eachlayer()** for example:  :D
http://LeafletJS.com/reference-1.3.0.html#map-eachlayer

cvalenzuela commented on src/providers/tile/Leaflet.js:
```diff
@@ -1,120 +1,109 @@
-    const cnvs = this.canvas.getContext('webgl') || this.canvas.getContext('2d');
+    L.overlay = () => new L.Layer.Overlay;
+
+    this.tiles && (this.tiles.options.opacity = this.options.opacity);
```
(just a note on replacing the short-circuit evaluations)

cvalenzuela commented on src/providers/tile/Leaflet.js:
```diff
@@ -1,120 +1,109 @@
-      };
-    })
+      const d = this.map.dragging._draggable._newPos;
+      d && (ctx.canvas.style.transform = `translate(${-d.x}px, ${-d.y}px)`);
```
same as above

cvalenzuela commented on src/providers/tile/Leaflet.js:
```diff
@@ -1,120 +1,109 @@
-      return this.map.getZoom();
-    }
-    return 0;
+    return this.ready? this.map.getZoom() : 0;
```
ternary

GoToLoop replied:
Well, that's 4 lines shortened into 1 short statement. RU sure do you prefer the 4-line fatty back here?

cvalenzuela replied:
I'm using the eslint with the airbnb style guide and that defaults to [no-ternary](https://eslint.org/docs/rules/no-ternary). I previously had everything written in ternary but at some point, it got messier to manage, so I switch everything to maintain consistency. Like [here](https://github.com/cvalenzuela/Mappa/blob/master/src/providers/static/Mapbox.js#L34). 

Still everything gets minified in prod so all good

GoToLoop replied:
Oh, that's a pity. I'm not advocating big multi-line ternaries; rather strategic 1s which is still clear & short like this 1.
However, if you've got such very strict eslint rule in place... well, let's go back to the ugly 4-line boilerplate.  :P

cvalenzuela replied:
cool! thanks

cvalenzuela reviewed (commented):
Looks good! I just left a few comments mostly on the style

cvalenzuela commented:
wow, this is amazing! thanks for the PR!

I left some comments, but regarding your points:

1. Perfect. Makes total sense.

2. Seems good to me, I just left some questions on the review.

3. I didn't know that! leaflet had a major released not so long ago and I haven't updated this repo to reflect that.

I will test it and let you know how it goes. I still haven't implemented all automatic tests so part of it will go manually.

Let me know your comments so we can move forward and merge

GoToLoop committed:
Style Revert

GoToLoop committed:
Ternary removed!

GoToLoop committed:
(1.3.x)

GoToLoop committed:
fromPointToLatLng()

Make **fromPointToLatLng()** return a stripped down raw { lat, lng } just like **fromLatLngToPixel()**.

GoToLoop committed:
Update Leaflet.js

GoToLoop committed:
canvasOverlay() -> onAdd() -> on()

GoToLoop committed:
Update Leaflet.js

GoToLoop committed:
this.map = this.tiles = null;

ChaimStanton commented:
I think this commit may change the id attribute of the layer from what it was previously as #mappa to #leaflet. See my issue #47 for more info.
